api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java

```diff
@@ -41,52 +49,73 @@
+  private static DbUser providedDbUser;
+  private static WorkbenchConfig providedWorkbenchConfig;
 
-  @Autowired private UserDao userDao;
-  @Mock private AdminActionHistoryDao adminActionHistoryDao;
-  @Autowired private UserDataUseAgreementDao userDataUseAgreementDao;
-  @Mock private FireCloudService fireCloudService;
-  @Mock private ComplianceService complianceService;
-  @Mock private DirectoryService directoryService;
+  @Autowired private FireCloudService mockFireCloudService;
+  @Autowired private ComplianceService mockComplianceService;
+  @Autowired private DirectoryService mockDirectoryService;
+  @Autowired private UserServiceAuditAdapter mockUserServiceAuditAdapter;
 
-  private UserService userService;
+  @Autowired private UserService userService;
+  @Autowired private UserDao userDao;
 
-  private DbUser testUser;
+  @TestConfiguration
+  @Import({UserService.class})
+  @MockBean({
+    AdminActionHistoryDao.class,
+    FireCloudService.class,
+    ComplianceService.class,
+    DirectoryService.class,
+    UserServiceAuditAdapter.class
+  })
+  static class Configuration {
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_SINGLETON)
+    Clock clock() {
+      return PROVIDED_CLOCK;
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    WorkbenchConfig getWorkbenchConfig() {
+      return providedWorkbenchConfig;
+    }
+
+    @Bean
+    Random getRandom() {
+      return new Random();
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    DbUser getDbUser() {
+      return providedDbUser;
+    }
+  }
 
   @Before
   public void setUp() {
-    CLOCK.setInstant(Instant.ofEpochMilli(TIMESTAMP_MSECS));
-    Provider<WorkbenchConfig> configProvider = Providers.of(WorkbenchConfig.createEmptyConfig());
-    testUser = insertUser(EMAIL_ADDRESS);
-
-    userService =
-        new UserService(
-            Providers.of(testUser),
-            userDao,
-            adminActionHistoryDao,
-            userDataUseAgreementDao,
-            CLOCK,
-            new Random(),
-            fireCloudService,
-            configProvider,
-            complianceService,
-            directoryService);
-  }
-
-  private DbUser insertUser(String email) {
     DbUser user = new DbUser();
-    user.setEmail(email);
-    user.setUserId(incrementedUserId);
-    incrementedUserId++;
+    user.setEmail(EMAIL_ADDRESS);
     userDao.save(user);
-    return user;
+    providedDbUser = user;
+
+    providedWorkbenchConfig = WorkbenchConfig.createEmptyConfig();
+
+    // Since we're injecting the same static instance of this FakeClock,
+    // increments and other mutations will carry across tests if we don't reset it here.
+    // I tried easier ways of ensuring this, like creating new instances for every test run,
+    // but it was injecting null clocks giving NPEs long after construction, and so far this
+    // is the only working approach I've seen.
+    PROVIDED_CLOCK.setInstant(START_INSTANT);
```

Comment:
if you make it a prototype and remove final, you should be able to reinitialize it here. Basically what we do for user and workbenchConfig above.

Reply:
Is that preferred? I kind of like this mental model, just from a general aversion to reassigning references where possible.